File: Desktop/Services/RssiLabEngine.cs
namespace KitHerramientas.Desktop.Services;

public enum LabSensitivity
{
    High,
    Normal,
    Low
}

public sealed record RssiBaseline(double Mean, double StdDev, int Samples);

public sealed record RssiLabReading(
    DateTimeOffset Timestamp,
    int Rssi,
    double Baseline,
    double Delta,
    double Score,
    string State,
    string? Marker = null)
{
    public string TimeText => Timestamp.LocalDateTime.ToString("HH:mm:ss");
    public string MarkerText => Marker ?? "";
}

public sealed class RssiLabEngine
{
    public RssiBaseline? Baseline { get; private set; }
    public LabSensitivity Sensitivity { get; set; } = LabSensitivity.Normal;
    private int? _previousRssi;

    public string SensitivityLabel => Sensitivity switch
    {
        LabSensitivity.High => "ALTA",
        LabSensitivity.Low => "BAJA",
        _ => "NORMAL"
    };

    public void ClearCalibration()
    {
        Baseline = null;
        _previousRssi = null;
    }

    public void ResetTracking() => _previousRssi = null;

    public RssiBaseline? Calibrate(IReadOnlyCollection<int> samples)
    {
        if (samples.Count < 5) return null;
        var mean = samples.Average();
        var variance = samples.Sum(v => Math.Pow(v - mean, 2)) / samples.Count;
        var std = Math.Max(1.0, Math.Sqrt(variance));
        Baseline = new RssiBaseline(mean, std, samples.Count);
        _previousRssi = samples.LastOrDefault();
        return Baseline;
    }

    public RssiLabReading? Evaluate(int rssi, string? marker = null, DateTimeOffset? timestamp = null)
    {
        var baseline = Baseline;
        if (baseline is null) return null;

        var delta = Math.Abs(rssi - baseline.Mean);
        var jump = _previousRssi is null ? 0.0 : Math.Abs(rssi - _previousRssi.Value);
        var noise = Math.Max(1.25, baseline.StdDev + 0.60);
        var rawScore = Math.Max(delta / noise, (jump / noise) * 0.90);
        var scale = Sensitivity switch
        {
            LabSensitivity.High => 0.78,
            LabSensitivity.Low => 1.30,
            _ => 1.0
        };
        var score = rawScore / scale;
        var state = score switch
        {
            < 1.40 => "ESTABLE",
            < 2.40 => "CAMBIO LEVE",
            < 3.60 => "CAMBIO FUERTE",
            _ => "VARIACIÓN COMPATIBLE CON MOVIMIENTO"
        };

        _previousRssi = rssi;
        return new RssiLabReading(timestamp ?? DateTimeOffset.Now, rssi, baseline.Mean, delta, score, state, marker);
    }
}

File: Android/app/src/main/java/com/max/kitherramientas/RssiLabEngine.kt
package com.max.kitherramientas

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sqrt

enum class LabSensitivity(val displayName: String, val scale: Double) {
    HIGH("ALTA", 0.78),
    NORMAL("NORMAL", 1.0),
    LOW("BAJA", 1.30)
}

data class RssiBaseline(val mean: Double, val stdDev: Double, val samples: Int)

data class RssiLabReading(
    val timestampMs: Long,
    val rssi: Int,
    val baseline: Double,
    val delta: Double,
    val score: Double,
    val state: String,
    val marker: String? = null
)

class RssiLabEngine {
    var baseline: RssiBaseline? = null
        private set

    var sensitivity: LabSensitivity = LabSensitivity.NORMAL
    private var previousRssi: Int? = null

    fun clearCalibration() {
        baseline = null
        previousRssi = null
    }

    fun resetTracking() {
        previousRssi = null
    }

    fun calibrate(samples: List<Int>): RssiBaseline? {
        if (samples.size < 5) return null
        val mean = samples.average()
        val variance = samples.sumOf { (it - mean) * (it - mean) } / samples.size
        val std = max(1.0, sqrt(variance))
        return RssiBaseline(mean, std, samples.size).also {
            baseline = it
            previousRssi = samples.lastOrNull()
        }
    }

    fun evaluate(rssi: Int, timestampMs: Long = System.currentTimeMillis(), marker: String? = null): RssiLabReading? {
        val base = baseline ?: return null
        val delta = abs(rssi - base.mean)
        val jump = previousRssi?.let { abs(rssi - it).toDouble() } ?: 0.0
        val noise = max(1.25, base.stdDev + 0.60)
        val rawScore = max(delta / noise, (jump / noise) * 0.90)
        val score = rawScore / sensitivity.scale
        val state = when {
            score < 1.40 -> "ESTABLE"
            score < 2.40 -> "CAMBIO LEVE"
            score < 3.60 -> "CAMBIO FUERTE"
            else -> "VARIACIÓN COMPATIBLE CON MOVIMIENTO"
        }
        previousRssi = rssi
        return RssiLabReading(timestampMs, rssi, base.mean, delta, score, state, marker)
    }
}
